Create 2022-12-08-common-kibana-searches-on-clouddotgov.md (#2218)
18F/cg-site: _kbarticles/2022-12-08-common-kibana-searches-on-clouddotgov.md

---
layout: post
title: "Common Kibana Searches on cloud.gov"
date: December 8, 2022
excerpt: This article documents a brief explaination of what Kibana does and common Kibana search queries on the cloud.gov platform.

---

## What does Kibana do?

> Application logs are visualized via Kibana, which has a [user guide](https://www.elastic.co/guide/en/kibana/current/index.html) that explains more about how to use it and customize your views.

Reviewer: ```suggestion
Kibana is a user interface that lets you search and visualize your application logs. Kibana has a [user guide](https://www.elastic.co/guide/en/kibana/current/index.html) that explains more about how to use it and to create custom visualizations.
```

## What are some types of logs?

> Cloud Foundry assigns a type to each log message depending on its origin. Application logs are assigned the APP log type.  HTTP requests being routed to an app will produce the RTR log type. The various types of logs are listed in the documentation [here](https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format).

Reviewer: ```suggestion
Cloud Foundry assigns a type to each log message depending on its origin. Application logs are assigned the APP log type.  HTTP requests being routed to an app will produce the RTR log type. The various types of logs are listed in the documentation [here](https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format).

Cloudfoundry logs are often translated into field names in Elasticsearch/Kibana using the log type as a prefix. For example, APP logs in Kibana include fields like `app.name` for the application name.
```

I think it's helpful to give some idea of how CloudFoundry log types are translated to searchable fields in Kibana

## How to chart the information:

> Spikes in traffic can be visualized by using the chart provided in Kibana. To find the specific information you are requesting you need to use the Discover tab to search RTR logs, and then create the chart based on the saved search. Use a count aggregation as the Y-Axis and a date histogram aggregation for the X-Axis.

Reviewer: I think it would be pretty helpful to create a screenshot showing how to configure this type of visualization. And then a screenshot of the end result. 

Use a test app to create the visualization and screenshot, not any customer app, so that we don't expose customer information

I can pair on this with you if you need help

Author: Mark, Thank you for your help with this! Added the images.

> ## Listed below are some common RTR search requests to input into Kibana:

Reviewer: ```suggestion
## Fields for router requests

Listed below are the explanations of some field names for router (RTR) logs:
```

---
layout: post

> title: "Common Kibana Searches on cloud.gov"

Reviewer: ```suggestion
title: "Understanding Kibana and how to visualize your application logs"
```

Let's make the title really specific to the customer value of the article. And I think this article is really valuable

date: December 8, 2022

> excerpt: This article documents a brief explaination of what Kibana does and common Kibana search queries on the cloud.gov platform.

Reviewer: ```suggestion
excerpt: This article briefly explains what Kibana does, what types of logs are available in Kibana, and how to create log visualizations in Kibana.
```

## How to chart the information:

Spikes in traffic can be visualized by using the chart provided in Kibana. To find the specific information you are requesting you need to use the Discover tab to search RTR logs, and then create the chart based on the saved search. Use a count aggregation as the Y-Axis and a date histogram aggregation for the X-Axis.

## Fields for router requests

Listed below are the explanations of some field names for router (RTR) logs:

* **rtr.app.id**: The application guid
* **rtr.hostname**: The domain/hostname the request was sent to (e.g test.app.cloud.gov)
* **rtr.http_user_agent**: What user agent the request came from (Chrome, Firefox, Curl, etc…)
* **rtr.path**: The specific url path that was requested (e.g. /my/test/page)
* **rtr.status**: Gives the status of the request (200, 404, etc…)
* **rtr.verb**: The type of request (POST, GET, etc...)
* **rtr.x_forwarded_for**: The IP address the request came from

> * **rtr.timestamp**: The time of the request in UTC

Reviewer: ```suggestion
* **rtr.timestamp**: The time of the request in UTC

The full list of fields available for router logs can be found in [our Elasticsearch field mapping configuration](https://github.com/cloud-gov/logsearch-for-cloudfoundry/blob/develop/jobs/elasticsearch-config-lfc/templates/component-index-mappings-app.json.erb#L38).
```

## What are some types of logs?

Cloud Foundry assigns a type to each log message depending on its origin. Application logs are assigned the APP log type.  HTTP requests being routed to an app will produce the RTR log type. The various types of logs are listed in the documentation [here](https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format).

Cloudfoundry logs are often translated into field names in Elasticsearch/Kibana using the log type as a prefix. For example, APP logs in Kibana include fields like `app.name` for the application name.

## How to visualize application traffic

Router log data can be used to create a visualization of your application traffic following the steps below. 

After you have logged into Kibana, click "Discover" in the left sidebar menu. Then, add filters and search terms to query for router logs as seen in the screenshot below. Please note that the filters shown here for a specific space and application are just an example. You might want to view logs for all requests to application in a given space, in which case you would not want a filter for `@cf.app`.  

> ![Screenshot of Kibana interface showing a query for router logs for a given CloudFoundry space and application](https://user-images.githubusercontent.com/104385372/207403558-68266274-af72-43e1-b869-c9ad1805adca.png)

Reviewer: ```suggestion

![Screenshot of Kibana interface showing a query for router logs for a given CloudFoundry space and application](https://user-images.githubusercontent.com/104385372/207403558-68266274-af72-43e1-b869-c9ad1805adca.png)
```

The next step is to visualize your search results based on a specific field. To visualize request logs over time, choose the `@timestamp` field from the left sidebar of "Available fields". Then, click "Visualize".

> ![Screenshot of Kibana interface showing the selection of @timestamp as field to use for visualizing request logs](https://user-images.githubusercontent.com/104385372/207403643-d2aa8bb8-702b-4aee-a028-baa7caf25fee.png)

Reviewer: ```suggestion

![Screenshot of Kibana interface showing the selection of @timestamp as field to use for visualizing request logs](https://user-images.githubusercontent.com/104385372/207403643-d2aa8bb8-702b-4aee-a028-baa7caf25fee.png)
```

By default, visualizing logs based on `@timestamp` will produce a histogram chart. To change the chart type to line, which might be more useful for this type of data, click the "Metrics & axes" link in the chart configuration panel on the right side of the screen. Then, under "Metrics" and "Count", select "Line" from the "Chart type" drop-down. Finally, click the "Update" button at the bottom right of the screen and the chart should update to a line chart.

> ![Screenshot of Kibana visualization interface showing a line chart of requests over time with a panel for chart configuration options visible on the right side of the screen. In the chart configuration options, "line" is selected as the chart type](https://user-images.githubusercontent.com/104385372/207403815-9f98ff91-6c1a-4f87-97ea-bd60e285ec49.png)

Reviewer: ```suggestion

![Screenshot of Kibana visualization interface showing a line chart of requests over time with a panel for chart configuration options visible on the right side of the screen. In the chart configuration options, "line" is selected as the chart type](https://user-images.githubusercontent.com/104385372/207403815-9f98ff91-6c1a-4f87-97ea-bd60e285ec49.png)
```

---
layout: post
title: "Understanding Kibana and how to visualize your application logs"
date: December 8, 2022
excerpt: This article briefly explains what Kibana does, what types of logs are available in Kibana, and how to create log visualizations in Kibana.

---

## What does Kibana do?

Kibana is a user interface that lets you search and visualize your application logs. Kibana has a [user guide](https://www.elastic.co/guide/en/kibana/current/index.html) that explains more about how to use it and to create custom visualizations.

## What are some types of logs?

Cloud Foundry assigns a type to each log message depending on its origin. Application logs are assigned the APP log type.  HTTP requests being routed to an app will produce the RTR log type. The various types of logs are listed in the documentation [here](https://docs.cloudfoundry.org/devguide/deploy-apps/streaming-logs.html#format).

Cloudfoundry logs are often translated into field names in Elasticsearch/Kibana using the log type as a prefix. For example, APP logs in Kibana include fields like `app.name` for the application name.

> ## How to chart the information:

Reviewer: ```suggestion
## How to visualize application traffic
```

> Spikes in traffic can be visualized by using the chart provided in Kibana. To find the specific information you are requesting you need to use the Discover tab to search RTR logs, and then create the chart based on the saved search. Use a count aggregation as the Y-Axis and a date histogram aggregation for the X-Axis.

Reviewer: ```suggestion
Router log data can be used to create a visualization of your application traffic following the steps below. 
```

> ## Discover Search

Reviewer: ```suggestion
After you have logged into Kibana, click "Discover" in the left sidebar menu. Then, add filters and search terms to query for router logs as seen in the screenshot below. Please note that the filters shown here for a specific space and application are just an example. You might want to view logs for all requests to application in a given space, in which case you would not want a filter for `@cf.app`.  
```

> ![Discover Search](https://user-images.githubusercontent.com/104385372/207403558-68266274-af72-43e1-b869-c9ad1805adca.png)

Reviewer: ```suggestion
![Screenshot of Kibana interface showing a query for router logs for a given CloudFoundry space and application](https://user-images.githubusercontent.com/104385372/207403558-68266274-af72-43e1-b869-c9ad1805adca.png)
```

> ## Add Timestamp Visualization

Reviewer: ```suggestion
The next step is to visualize your search results based on a specific field. To visualize request logs over time, choose the `@timestamp` field from the left sidebar of "Available fields". Then, click "Visualize".
```

> ![Add Timestamp Visualization](https://user-images.githubusercontent.com/104385372/207403643-d2aa8bb8-702b-4aee-a028-baa7caf25fee.png)

Reviewer: ```suggestion
![Screenshot of Kibana interface showing the selection of @timestamp as field to use for visualizing request logs](https://user-images.githubusercontent.com/104385372/207403643-d2aa8bb8-702b-4aee-a028-baa7caf25fee.png)
```

Reviewer: [This text in the brackets is known as "alt" text and it is read out loud by screen readers for blind users.](https://moz.com/learn/seo/alt-text) So it's important the text be descriptive of what the image actually shows.

Author: Thank you, sorry about that, was unaware.

> ## Add Date Histogram Aggregation for X-Axis
> ![Add Date Histogram Aggregation for X-Axis](https://user-images.githubusercontent.com/104385372/207403718-60e70970-7dd4-452f-9d57-8c2230a5c450.png)

Reviewer: ```suggestion
```

This screenshot is misleading because it shows a line chart when it fact a histogram is produced by default. I think it might be easier to just exclude it.

Author: I agree

> ## Add Count Aggregation for Y-Axis/Line Chart Visualization

Reviewer: ```suggestion
By default, visualizing logs based on `@timestamp` will produce a histogram chart. To change the chart type to line, which might be more useful for this type of data, click the "Metrics & axes" link in the chart configuration panel on the right side of the screen. Then, under "Metrics" and "Count", select "Line" from the "Chart type" drop-down. Finally, click the "Update" button at the bottom right of the screen and the chart should update to a line chart.
```

> ![Add Count Aggregation for Y-Axis / Line Chart Visualization](https://user-images.githubusercontent.com/104385372/207403815-9f98ff91-6c1a-4f87-97ea-bd60e285ec49.png)

Reviewer: ```suggestion
![Screenshot of Kibana visualization interface showing a line chart of requests over time with a panel for chart configuration options visible on the right side of the screen. In the chart configuration options, "line" is selected as the chart type](https://user-images.githubusercontent.com/104385372/207403815-9f98ff91-6c1a-4f87-97ea-bd60e285ec49.png)
```
